### PR TITLE
O3-1431: Create an appointment.

### PR DESCRIPTION
**Pull request by hadijahkyampeire**
## Requirements

- [X] This PR has a title that briefly describes the work done, including the ticket number if there is a ticket.
- [X] My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide).
- [X] I checked for feature overlap with [**existing widgets**](https://om.rs/directory).

## Summary

- Implement creating a new appointment
- Some few fixes

## Screenshots

https://user-images.githubusercontent.com/30952856/184040105-879147b4-91a0-46f8-98c4-c32e381d4345.mov

- No patient was found with that search name.

Screenshot 2022-08-05 at 02 21 49

## Related Issue

[Create an Appointment
](https://issues.openmrs.org/browse/O3-1431)

## Other

*None.*

### CODE REVIEW

**Comment by CynthiaKamau** [comment]
Great work! Kindly resolve the conflicts

**Comment by hadijahkyampeire** [comment]
> Great work! Kindly resolve the conflicts

I sure will, thanks @CynthiaKamau .

**Comment by hadijahkyampeire** [comment]
@CynthiaKamau I still have to do some more refactoring, maybe make one form for editing and creating, we can decide to do this in a follow up PR or this one.

**Comment by CynthiaKamau** [comment]
> @CynthiaKamau I still have to do some more refactoring, maybe make one form for editing and creating, we can decide to do this in a follow up PR or this one.

Yes, reusung the appointment form would be easier

**Comment by hadijahkyampeire** [comment]
@CynthiaKamau I made the changes we talked about and also fixed the patient-header issue, now it can be viewed while creating or editing an appointment

**Review by CynthiaKamau — APPROVED** [review]

**Comment by CynthiaKamau** [comment]
Kindly fix the conflicts

File: packages/esm-appointments-app/src/appoinments-tabs/appointments-base-table.component.tsx [hunk 1]
```diff
@@ -31,7 +31,7 @@ import styles from './appointments-base-table.scss';
 import { MappedAppointment } from '../types';
 import { launchOverlay } from '../hooks/useOverlay';
 import AppointmentDetails from '../appointment-details/appointment-details.component';
-import AppointmentForm from '../appointment-forms/appointment-form.component';
+import AppointmentForm from '../appointment-forms/edit-appointment-form.component';
```
Comment (ZacButko) [inline packages/esm-appointments-app/src/appoinments-tabs/appointments-base-table.component.tsx:34]: Just noticing the file name change.. is AppointmentForm only for edit? or can it be used for adding forms also?

File: packages/esm-appointments-app/src/appointment-forms/appointment-forms.resource.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  return openmrsFetch(apiUrl, {
+    signal: abortController.signal,
+  });
+}
+
+export function useProviders() {
+  const { data, error } = useSWR<{ data: { results: Array<Provider> } }, Error>(`/ws/rest/v1/provider`, openmrsFetch);
+
+  return {
+    data: data ? data.data.results : null,
```
Comment (ZacButko) [inline packages/esm-appointments-app/src/appointment-forms/appointment-forms.resource.ts:55]: I would add optional chaining here just to be safe

File: packages/esm-appointments-app/src/appointment-forms/appointment-forms.resource.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    method: 'POST',
+    signal: abortController.signal,
+    headers: {
+      'Content-Type': 'application/json',
+    },
+    body: appointment,
+  });
+}
+
+export const weekdays = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
```
Comment (ZacButko) [inline packages/esm-appointments-app/src/appointment-forms/appointment-forms.resource.ts:72]: Is this being used anywhere? In `create-appointment-form.component` you get `daysOfTheWeek` from config instead...

Comment (ZacButko) [reply packages/esm-appointments-app/src/appointment-forms/appointment-forms.resource.ts:72]: (I would delete it if not being used by anything)

File: packages/esm-appointments-app/src/appointment-forms/create-appointment-form.component.tsx [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+  const { t } = useTranslation();
+  const isTablet = useLayoutType() === 'tablet';
+  const { mutate } = useSWRConfig();
+  const locations = useLocations();
+  const session = useSession();
+  const [appointmentNote, setAppointmentNote] = useState('');
+  const [selectedProvider, setSelectedProvider] = useState('');
+  const [selectedService, setSelectedService] = useState('');
+  const [startDate, setStartDate] = useState(new Date());
+  const [startTime, setStartTime] = useState(dayjs(new Date()).format('hh:mm'));
```
Comment (ZacButko) [inline packages/esm-appointments-app/src/appointment-forms/create-appointment-form.component.tsx:63]: Is it possible to change this to be just one `Date` object? UTC is much more reliable than recording date and time separately, especially looking to future proof against timezones

File: packages/esm-appointments-app/src/appointment-forms/create-appointment-form.component.tsx [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+                datePickerType="single"
+                dateFormat="d/m/Y"
+                light={isTablet}
+                value={startDate}
+                onChange={([date]) => setStartDate(date)}>
+                <DatePickerInput
+                  id="datePickerInput"
+                  labelText={t('date', 'Date')}
+                  style={{ width: '100%' }}
+                  placeholder="dd/mm/yyyy"
```
Comment (ZacButko) [inline packages/esm-appointments-app/src/appointment-forms/create-appointment-form.component.tsx:187]: This time format should be implementation specific

File: packages/esm-appointments-app/src/appointment-forms/create-appointment-form.component.tsx [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+              aria-label="toggle button"
+              id="allDay"
+              toggled={allDayOn}
+              labelText="All Day"
+              onToggle={() => setAllDayOn(!allDayOn)}
+            />
+            <div className={styles.datePickerControl}>
+              <DatePicker
+                datePickerType="single"
+                dateFormat="d/m/Y"
```
Comment (ZacButko) [inline packages/esm-appointments-app/src/appointment-forms/create-appointment-form.component.tsx:179]: This time format should be implementation specific

File: packages/esm-appointments-app/src/appointment-forms/create-appointment-form.scss [hunk 7]
```diff
@@ -0,0 +1,145 @@
+@import '~@openmrs/esm-styleguide/src/vars';
+@import '~carbon-components/src/globals/scss/vars';
+@import '~carbon-components/src/globals/scss/mixins';
+@import '~carbon-components/scss/globals/scss/vendor/@carbon/elements/scss/type/styles';
+@import '~carbon-components/src/globals/scss/vendor/@carbon/layout/scss/generated/spacing';
+
+// TO DO Move this styles to style - guide
+// https://github.com/openmrs/openmrs-esm-core/blob/master/packages/framework/esm-styleguide/src/_vars.scss
+$color-blue-30: #a6c8ff;
```
Comment (ZacButko) [inline packages/esm-appointments-app/src/appointment-forms/create-appointment-form.scss:9]: `$carbon--blue-30` should exist in the above imports

File: packages/esm-appointments-app/src/appointment-forms/create-appointment-form.scss [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+.productiveHeading02 {
+  color: $color-gray-70;
+  @include carbon--type-style('productive-heading-02');
+}
+
+$openmrs-background-grey: #ededed;
```
Comment (ZacButko) [inline packages/esm-appointments-app/src/appointment-forms/create-appointment-form.scss:16]: shouldn't need to declare this

File: packages/esm-appointments-app/translations/en.json [hunk 9]
```diff
@@ -31,7 +38,7 @@
   "dateTime": "Date & time",
   "day": "Day",
   "discard": "Discard",
-  "editAppointment": "Edit Appointment",
+  "editAppointment": "EditAppointment",
```
Comment (ZacButko) [inline packages/esm-appointments-app/translations/en.json:41]: should be two separate words

File: packages/esm-appointments-app/src/appointment-forms/create-appointment-form.component.tsx (hunk 4 above)
Comment (hadijahkyampeire) [reply packages/esm-appointments-app/src/appointment-forms/create-appointment-form.component.tsx:63]: Yah, I was thinking of that too, I will add it to follow up PR

File: packages/esm-appointments-app/src/appointment-forms/create-appointment-form.scss (hunk 7 above)
Comment (hadijahkyampeire) [reply packages/esm-appointments-app/src/appointment-forms/create-appointment-form.scss:9]: I thought so too but it wasn't there in the [var.scss](https://github.com/openmrs/openmrs-esm-core/blob/master/packages/framework/esm-styleguide/src/_vars.scss)

File: packages/esm-appointments-app/src/appointment-forms/create-appointment-form.component.tsx (hunk 6 above)
Comment (hadijahkyampeire) [reply packages/esm-appointments-app/src/appointment-forms/create-appointment-form.component.tsx:179]: I don't understand this comment @ZacButko

**Comment by hadijahkyampeire** [comment]
@ZacButko I have noted a couple of great points from your review that I need to address i.e the Date UTC issue and making the form reusable, I request to have that in a follow-up PR since this PR has been already crowded and confusing.

File: packages/esm-appointments-app/src/appointment-forms/create-appointment-form.component.tsx (hunk 6 above)
Comment (ZacButko) [reply packages/esm-appointments-app/src/appointment-forms/create-appointment-form.component.tsx:175]: Different countries will format a date like d/m/y while others use m/d/y and still others use y/m/d. It should be enabled for an implementation to specify their preferred date format

File: packages/esm-appointments-app/src/appointment-forms/create-appointment-form.scss (hunk 7 above)
Comment (ZacButko) [reply packages/esm-appointments-app/src/appointment-forms/create-appointment-form.scss:9]: With the merge to 4.0 and carbon v11 it should exist in @carbon/colors. Let me double check the new naming for this color...

**Pull request merged**